### PR TITLE
Version upgrades

### PR DESCRIPTION
**Pull request by TswusayonaTsamsiyu**
Upgraded Node version and all the packages.
Then fixed configuration issues with Webpack, Babel, etc.
Removed some deprecated packages and replaced some with new equivalents.
Fixed a lot of breaking changes and compatibility issues of different packages.
Upgraded to the latest Material UI and modified practically all of the components to work again.
In many places I've also fixed bugs and improved the UI.
Oh and I also had to add the entire source of the saga-reducer-factory package in order to fix a compatibility issue in its core. We should open a PR to the original and then remove it.

It is recommended you look at per-commit diffs, as this is a very big change.
It will still probably look a little messy, so just ask me if anything's unclear.
Note that I am using the Prettier code formatter, so some files are significantly changed because of that.

### CODE REVIEW

File: src/saga-reducers-factory-patch/ActionCreatorHelper.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import { createAction } from 'redux-actions';
+import _                from 'lodash';
+
+// Expect a string in the form of 'ONE_TWO_THREE' and return 'oneTwoThree'
+const camelize = str => str.toLowerCase().replace(/_\w/g, cap => cap[1].toUpperCase());
```
Comment (sOfekS) [inline src/saga-reducers-factory-patch/ActionCreatorHelper.js:5]: toCamelCase would be clearer and more conventional.  Also pretty sure lodash has that function already just fyi

File: src/saga-reducers-factory-patch/ActionCreatorHelper.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// Expect an array of strings and return { ONE : 'ONE', TWO : 'TWO', THREE : 'THREE' }
+export const createTypes = (strings, prefix) => strings.reduce((rc, str) => {
+    let actionStr = null;
+    prefix = prefix || '';
+
+    if (typeof (str) === 'string') {
+        actionStr = str;
+    } else {
+        actionStr = str.type;
+    }
```
Comment (sOfekS) [inline src/saga-reducers-factory-patch/ActionCreatorHelper.js:16]:
```
const actionStr = (typeof (str) === 'string') ? str : str.type;
prefix = prefix || '';
```

File: src/saga-reducers-factory-patch/ActionCreatorHelper.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    let actionPropTypes = null;
+    prefix = prefix || '';
+
+    if (typeof (str) === 'string') {
+        actionStr = str;
+        actionPropTypes = null;
+    } else {
+        actionStr = str.type;
+        actionPropTypes = str.propTypes;
+    }
```
Comment (sOfekS) [inline src/saga-reducers-factory-patch/ActionCreatorHelper.js:34]: Same as before

File: src/components/CriteriaCreation/NumericCreation.js [hunk 4]
```diff
@@ -1,61 +1,76 @@
-import React, {Component} from 'react';
-import Subheader from 'material-ui/Subheader';
-import TextBox from '../FormComponents/TextBox';
-import {List, ListItem} from 'material-ui/List';
+import _ from "lodash";
+import PropTypes from "prop-types";
+import React, { Component } from "react";
+import { Typography, List, ListItem, InputAdornment, ListSubheader, withTheme } from "@material-ui/core";
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/CriteriaCreation/NumericCreation.js:4]: I believe this only affects dev mode, no? to make the bundling faster.
Webpack should automatically perform tree-shaking in production builds. At least that's my understanding.

File: src/components/CriteriaCreation/BooleanCreation.js [hunk 5]
```diff
@@ -1,19 +1,16 @@
...
-            <div className='boolean-creation'>
-                <Subheader>
-                    True / False Creation
-                </Subheader>
-                <div>
-                    <TextBox name='answers.true' placeholder='True answer'/>
-                    <TextBox name='answers.false' placeholder='False answer'/>
-                </div>
+
+export default function BooleanCreation() {
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/CriteriaCreation/BooleanCreation.js:6]: Hmm just wanted to give it a name for clarity. Anyway it doesn't really affect anything...

File: src/components/AddLeaderClinicModal.js [hunk 6]
```diff
@@ -41,78 +36,68 @@ class AddLeaderClinicModal extends Component {
...
 function mapDispatchToProps(dispatch) {
-    return bindActionCreators({
-        create: actions.create
-    }, dispatch);
+    return bindActionCreators(
+        {
+            create: actions.create,
+        },
+        dispatch
+    );
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/AddLeaderClinicModal.js:100]: It's the autoformating done by Prettier. I don't like it either, but it wasn't prioritized enough for me to tinker with it.

File: src/components/EditResonator/Steps/Criteria.js [hunk 7]
```diff
@@ -40,6 +40,7 @@ class EditResonatorCriteria extends Component {
     }
 
     render() {
+        const selectedCriteria = this.props.formData.criteria || [];
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/EditResonator/Steps/Criteria.js:43]: It's a leftover from when I enforced choosing criteria. Removed it.

File: src/components/FollowerResonators.js [hunk 8]
```diff
@@ -51,142 +45,102 @@ class FollowerResonators extends Component {
...
         return {
-            left: [
-            ],
-            right: [
-                <MoreOptionsMenu
-                    multiple
-                    value={moreOptions}
-                >
-                    <MenuItem onTouchTap={() => this.toggleShowInactive()} primaryText={text} value='showFrozen' />
+            left: <Typography variant="h6">{`${this.props.follower.user.name}'s Resonators`}</Typography>,
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/FollowerResonators.js:83]: I haven't rebuilt this with the recent changes for the preview I'm exposing to you. I'll do so soon once I finish with some extra fixes.

File: src/saga-reducers-factory-patch/ActionCreatorHelper.js (hunk 2 above)
Comment (TswusayonaTsamsiyu) [reply src/saga-reducers-factory-patch/ActionCreatorHelper.js:16]: `saga-reducers-factory` is an external package I copied into our code-base in order to fix compatibility issues it had with the new `redux-saga` version.
I'd rather not waste time on external dependency code :)
What we should do is open a PR to the original package. I believe @ItamarGronich knows the author, so we can speak to him.
This goes for all comments regarding `saga-reducers-factory`.

File: src/components/EditFollowerModal.js [hunk 9]
```diff
@@ -43,146 +39,120 @@ class EditFollowerModal extends Component {
...
     let clinics = state.clinics.clinics;
-    let current_clinic_id =  state.leaders.leaders.current_clinic_id;
+    let current_clinic_id = state.leaders.leaders.current_clinic_id;
 
     let ret = {
         follower,
         clinics,
-        editMode
+        editMode,
     };
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/EditFollowerModal.js:133]: Changed :)

File: src/components/EditResonator/Steps/activation.js [hunk 10]
```diff
@@ -6,19 +6,14 @@ import ToggleField from '../../FormComponents/ToggleField';
 class EditResonatorActivation extends Component {
     render() {
         return (
-            <div>
+            <div style={{display: "flex", flexDirection: "column"}}>
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/EditResonator/Steps/activation.js:9]: With class components one should use `withStyles`.
I'll add it to my tasks, but it seems to me to be of lower priority, and something we can do on-the-go.

File: src/components/EditResonator/Steps/schedule.js [hunk 11]
```diff
@@ -1,38 +1,37 @@
...
 
     constructor() {
         super();
 
         this.handleCreate = this.handleCreate.bind(this);
         this.handleUpdate = this.handleUpdate.bind(this);
         this.handleSelectTime = this.handleSelectTime.bind(this);
+
+        this.state = {
+            bla: new Date(),
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/EditResonator/Steps/schedule.js:29]: My bad. Removed it.

File: src/components/FollowerResonators.js [hunk 12]
```diff
@@ -51,142 +45,101 @@ class FollowerResonators extends Component {
...
-            <div className='row'>
-                <div className='name col-lg-10 col-sm-9 col-xs-6' style={{
-                    display: 'flex',
-                    width: '100%',
-                    background: 'rgb(187, 187, 187)'
-                }}>
-                    <b><p>Disabled Resonators</p></b>
+                        margin: "0 15px",
+                        textAlign: dir === "rtl" ? "right" : "left",
+                        color: resonator.pop_email === false ? "grey" : "",
```
[the remark this answers is not in the document]
Comment (TswusayonaTsamsiyu) [reply src/components/FollowerResonators.js:55]: Fixed!

**Pull request merged**